File: onstock/internal/httpapi/tenant.go
package httpapi

import (
	"net/http"
	"strings"

	"ondigital.hn/tenant"
)

func (a *API) getTenant(w http.ResponseWriter, r *http.Request) {
	view, err := a.st.PublicTenantView()
	if err != nil {
		writeErr(w, err)
		return
	}
	writeJSON(w, http.StatusOK, view)
}

type putTenantBody struct {
	Plan     string `json:"plan"`
	TenantID string `json:"tenant_id"`
	Modules  string `json:"modules"` // comma-separated
}

func (a *API) putTenant(w http.ResponseWriter, r *http.Request) {
	body, err := decode[putTenantBody](r)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "JSON inválido"})
		return
	}
	patch := map[string]string{}
	if body.Plan != "" {
		p, err := tenant.ParsePlan(body.Plan)
		if err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
			return
		}
		patch["plan"] = string(p)
	}
	if strings.TrimSpace(body.TenantID) != "" {
		patch["tenant_id"] = strings.TrimSpace(body.TenantID)
	}
	if body.Modules != "" {
		patch["modules"] = body.Modules
	}
	if len(patch) == 0 {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "nada que actualizar"})
		return
	}
	if err := a.st.SetSettings(patch); err != nil {
		writeErr(w, err)
		return
	}
	view, err := a.st.PublicTenantView()
	if err != nil {
		writeErr(w, err)
		return
	}
	writeJSON(w, http.StatusOK, view)
}

File: onstock/server/api/tenant.js
// Traducción de internal/httpapi/{tenant,modules}.go.
import { writeJSON, writeErr, readJSON, comoMapaDeGo } from './helpers.js';
import { parsePlan } from '../store/tenant.js';

export function getTenant(a, ctx) {
  writeJSON(ctx.res, 200, a.st.publicTenantView());
}

export async function putTenant(a, ctx) {
  let body;
  try {
    body = await readJSON(ctx.req);
  } catch {
    writeJSON(ctx.res, 400, { error: 'JSON inválido' });
    return;
  }
  const patch = {};
  if (body.plan) {
    let p;
    try {
      p = parsePlan(body.plan);
    } catch (err) {
      writeJSON(ctx.res, 400, { error: err.message });
      return;
    }
    patch.plan = p;
  }
  if (String(body.tenant_id ?? '').trim() !== '') patch.tenant_id = String(body.tenant_id).trim();
  if (body.modules) patch.modules = body.modules;
  if (Object.keys(patch).length === 0) {
    writeJSON(ctx.res, 400, { error: 'nada que actualizar' });
    return;
  }
  try {
    a.st.setSettings(patch);
  } catch (err) {
    writeErr(ctx.res, err);
    return;
  }
  writeJSON(ctx.res, 200, a.st.publicTenantView());
}

export function getModules(a, ctx) {
  if (!a.catalog) {
    writeJSON(ctx.res, 200, comoMapaDeGo({ modules: [], count: 0 }));
    return;
  }
  const infos = a.catalog.infos();
  writeJSON(ctx.res, 200, comoMapaDeGo({ modules: infos, count: infos.length }));
}
